addons/sale_margin/models/sale_order.py

```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from openerp import api, fields, models
+import openerp.addons.decimal_precision as dp
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    margin = fields.Float(compute='_product_margin', digits=dp.get_precision('Product Price'), store=True)
+    purchase_price = fields.Float(string='Cost Price', digits=dp.get_precision('Product Price'))
+
+    @api.multi
+    @api.onchange('product_id', 'product_uom_qty')
+    def product_id_change_margin(self):
+        for line in self:
+            if line.order_id.pricelist_id:
+                frm_cur = self.env.user.company_id.currency_id
+                to_cur = line.order_id.pricelist_id.currency_id
+                purchase_price = line.product_id.standard_price
+                if line.product_uom != line.product_id.uom_id:
+                    purchase_price = self.env['product.uom']._compute_price(line.product_id.uom_id.id, purchase_price, to_uom_id=line.product_uom.id)
+                ctx = self.env.context.copy()
+                ctx['date'] = line.order_id.date_order
+                price = frm_cur.with_context(ctx).compute(purchase_price, to_cur, round=False)
```

Comment:
No need to copy the context because it is used once. Just use: frm_cur.with_context(date=line.order_id.date_order).compute(purchase_price, to_cur, round=False) .